### PR TITLE
Remove JavaScript and CSS type attributes for HTML5 validity 

### LINKED ISSUE #339: Type attributes are added to javascript and CSS, this is not needed in HTML5
JavaScript and CSS resources queued in the current code includes a `type` attribute which is not necessary in HTML5. Themes should be able to direct core to not include this attribute when inserting JavaScript and CSS into the main theme. A WordPress ticked and patch exists for the same issue, it should be considered here too.

https://core.trac.wordpress.org/ticket/42804

## Expected behavior
JavaScript and CSS should be queued without the `type` attribute in HTML5 themes.

## Current behavior
JavaScript and CSS are currently queued with the `type` attribute in all circumstances unless hacked around.

## Possible solution
See ticket linked above with a proposed and patched solution.

## Steps to reproduce (for bugs)

1. Set up a vanilla install of ClassicPress
2. Validate the source on a HTML5 checker, note the warnings relating to the `type` attribute.

### PR DESCRIPTION
**Pull request by mattyrob**
## Description
This is a back port of upstream fixes over 5 commits:
https://core.trac.wordpress.org/changeset/46164
https://core.trac.wordpress.org/changeset/46165
https://core.trac.wordpress.org/changeset/46170
https://core.trac.wordpress.org/changeset/46171
https://core.trac.wordpress.org/changeset/46287

Hopefully I've applied the fixes correctly - there were lots of conflicts to fix in the 46165 changset

## Motivation and context
Fix #339 

## How has this been tested?
Unit tests included in patch

## Types of changes
- Bug fix

### CODE REVIEW

**Comment by bahiirwa** [comment]
Line 547 in `src/wp-includes/class.wp-scripts.php` has a git merge resolve error. `<<<< comments`left

**Comment by mattyrob** [comment]
@bahiirwa 

Thanks you for spotting that - I hope it's the only one. Update applied.

**Comment by bahiirwa** [comment]
That’s what I saw on first glance. Is there are difference between this PR and #481 ?

**Comment by mattyrob** [comment]
@bahiirwa

This PR is more extensive than mine and is a backport of a WordPress fix rather than my attempt from scratch.

**Comment by bahiirwa** [comment]
My real question now is do we merge both or do we close the earlier one?

**Comment by ginsterbusch** [comment]
So what happens if somebody uses XHTML oder XHTML5?

cu, w0lf.

**Comment by mattyrob** [comment]
@ginsterbusch

As far as I can tell in my reading XHTML currently stands officially at version 2.0 from Dec 2010 so HTML5 should be encouraged.

HTML5 and XHTML5 are identical in vocabularies but different in parsing. Further it is XHTML is an XML serialisation of HTML so they should be compatible but the priority at W3C appears to be HTML5.

**Comment by nylen** [comment]
I don't think XHTML[5] is relevant to this PR. Even if those specs do have something to say about a `type` attribute for script tags, themes that are still using XHTML would just not declare support for HTML5.

The changes to Twenty Seventeen look good. The changes to Twenty Fifteen and Sixteen appear to include too much code.

The commit message for changeset 46165 was also lost (not preserved in the required format), probably due to this part of the [backporting instructions](https://github.com/ClassicPress/ClassicPress/blob/develop/.github/CONTRIBUTING.md#backporting-changes-from-wordpress) being missed here:

> [The backport script] uses a standardized format for commit messages which makes it possible for us to track which WordPress changes we've already included. If there are merge conflicts for your changes, be sure to use `git cherry-pick --continue` after resolving them, because this will preserve the commit message in the required format.

**Comment by mattyrob** [comment]
This patch comes with unit tests.

But also note that on a vanilla install of `develop` ClassicPress the term `text/css` appears six times in the page source on the home page and `text/javascript` appears 10 times.

With the patch in place the number of appearances of `text/css` is zero and `text/javascript` occurs once as part of the source of the enqueued Emoji scripts, but never as a type attribute.

**Comment by mattyrob** [comment]
In reviewing this patch I have realised that it introduces 2 new functions to the `WP_Scripts` class that hand setting and printing of translation strings to JS files.

We need a decision here to either:

- Backport these changes and update the patch to correct the inline documentation
- Strip out these 2 functions and any calls to them

**Comment by striebwj** [comment]
> Backport these changes and update the patch to correct the inline documentation
> Strip out these 2 functions and any calls to them

Happy for your's and @bahiirwa's input on these options.

**Comment by mattyrob** [comment]
> > Backport these changes and update the patch to correct the inline documentation
> > Strip out these 2 functions and any calls to them
> 
> Happy for your's and @bahiirwa's input on these options.

I have looked a bit into this and these functions are used extensively upstream in the `script-loader.php` file to `set_translations` and then just once to output those translations - I can only premise it was a necessary addition for ensiring the block editor (and any plugins extending it) were i18n friendly.

I'm of the opinion that we can drop these from CP.

**Comment by mattyrob** [comment]
Broken unit tests was a missed prune to the updated code on my part - updated PR to follow.

File: src/wp-includes/formatting.php [hunk 1]
```diff
@@ -5082,6 +5084,9 @@ function _print_emoji_detection_script() {
 		'svgExt' => apply_filters( 'emoji_svg_ext', '.svg' ),
 	);
 
+	$version   = 'ver=' . get_bloginfo( 'version' );
```
Comment (striebwj) [inline src/wp-includes/formatting.php:5087]: Where does this get used?

Comment (mattyrob) [reply src/wp-includes/formatting.php]: It doesn't - good spot. I suspect it's a backport artefact, removed in updated PR.

**Review by striebwj — APPROVED** [review]

File: src/wp-includes/script-loader.php [hunk 2]
```diff
@@ -1520,11 +1526,15 @@ function _print_styles() {
 		$concat = str_split( $concat, 128 );
 		$concat = 'load%5B%5D=' . implode( '&load%5B%5D=', $concat );
 
-		$href = $wp_styles->base_url . "/wp-admin/load-styles.php?c={$zip}&dir={$dir}&" . $concat . '&ver=' . $ver;
-		echo "<link rel='stylesheet' href='" . esc_attr($href) . "' type='text/css' media='all' />\n";
+		foreach ( $concat as $key => $chunk ) {
```
Comment (joyously) [inline src/wp-includes/script-loader.php:1529]: Remove this foreach loop.

File: src/wp-includes/script-loader.php [hunk 3]
```diff
@@ -1520,11 +1526,15 @@ function _print_styles() {
...
+			$concatenated .= "&load%5Bchunk_{$key}%5D={$chunk}";
+		}
+
+		$href = $wp_styles->base_url . "/wp-admin/load-styles.php?c={$zip}&dir={$dir}" . $concatenated . '&ver=' . $ver;
```
Comment (joyously) [inline src/wp-includes/script-loader.php:1533]: There is a missing `&` after the dir variable. 
Change $concatenated to $concat.

**Pull request merged**